runtimes/starlarkrt/runtime/builtins.go

```diff
@@ -20,17 +22,69 @@ var builtins = map[string]func(thread *starlark.Thread, fn *starlark.Builtin, ar
 
 		value, err := starlark.Call(thread, fn, args[1:], kwargs)
 		if err != nil {
+			if perr, ok := sdktypes.FromError(err); ok {
+				v := perr.Value()
+				vctx := values.Context{}
+				if slv, cerr := vctx.ToStarlarkValue(v); cerr == nil {
+					return starlark.Tuple{starlark.None, slv}, nil
+				}
+			}
+
 			return starlark.Tuple{starlark.None, starlark.String(err.Error())}, nil
 		}
 
 		return starlark.Tuple{value, starlark.None}, nil
 	},
 	"fail": func(thread *starlark.Thread, bi *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
-		var msg string
-		if err := starlark.UnpackArgs(bi.Name(), args, kwargs, "msg", &msg); err != nil {
-			return nil, err
+		if len(args) > 0 && len(kwargs) > 0 {
+			return nil, fmt.Errorf("fail: cannot specify both positional and keyword arguments")
 		}
-		return nil, errors.New(msg)
+
+		vctx := values.Context{}
+
+		if len(args) == 1 {
+			v, err := vctx.FromStarlarkValue(args[0])
+			if err != nil {
+				return nil, fmt.Errorf("cannot convert value from starlark: %w", err)
+			}
+			return nil, sdktypes.NewProgramError(v, nil, nil).ToError()
+		} else if len(args) > 1 {
```

Comment:
since you are returning from the if condition, please remove the else
it makes code less indented and more readable

Reply:
done